source/common/router/router.cc

```diff
@@ -280,6 +282,38 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::HeaderMap& headers, bool e
   // that get handled by earlier filters.
   config_.stats_.rq_total_.inc();
 
+  // Process metadata...
+  bool do_not_forward = false;
+  const auto& metadata = callbacks_->streamInfo().dynamicMetadata().filter_metadata();
+  const auto fields_it = metadata.find(Extensions::HttpFilters::HttpFilterNames::get().Router);
```

Comment:
Do we have any standard way of documenting metadata and its effects? This is basically a form of API that we want people to know about.

Reply:
Stepping back, I'm not super crazy about using metadata in this way:
1) It's not very performant.
2) As you point out, it's a pseudo API that is quite confusing.

As I see it there are some other options:
1) We could have some explicit attributes that are generally useful accessed through stream info (though this is a bit strange as it's not really info)
2) Have state like this hanging off of the filter callbacks, so that filter can set/interrogate this these boolean type settings.

Thoughts?

Reply:
Yeah, I'm sort of using metadata here as a cheap stand-in for what I imagine we might want to build in the 3.x API world - having the unified Tap filter style matchers accessible from StreamInfo (the "rest of the owl" from my [stream tagging](https://docs.google.com/document/d/14CAt13uAlrCPPAh3dL06JL0FrKRh8VKJfjCg6HV7p3c/edit) doc).

I agree it feels a bit janky. But the thing I like about this approach in the short term is that it doesn't require building any new configuration mechanism that duplicates the other various matcher implementations. It just assumes that you have a way of attaching metadata to a stream, such as the headers-to-metadata filter. If we wanted to do something different (e.g. new boolean attributes on StreamInfo or the filter callbacks), we'd need some other way to configure/set them.

Let's chat if you have a few minutes today.

Reply:
> 1. We could have some explicit attributes that are generally useful accessed through stream info (though this is a bit strange as it's not really info)
> 2. Have state like this hanging off of the filter callbacks, so that filter can set/interrogate this these boolean type settings.

If we're taking a step back, I'd like to create a taxonomy of the various filter communication and sharing mechanisms such as what was done in https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/data_sharing_between_filters#typed-metadata.

In addition to ^^, we have `StreamInfo`, and now we have a fourth mechanism? I'm running out of fingers to count with :) Is there a way to get some clear picture on the variety of "stream metadata" mechanisms and when each is appropriate?

Reply:
@mattklein123, have a look at the latest. I implemented `StreamDebugInfo` as a sort of sibling of `StreamInfo`. Is this what you were picturing, directionally?

Reply:
Yeah this is exactly what I hand in mind. Looks like a great start.

/wait